File: Examples/VB/Light/Light.vb
Imports StateForge.StateMachine

Namespace StateForge.Examples.Light

	Public Interface ILightActuator
		Sub DoOn()
		Sub DoOff()
	End Interface

	Public Class LightActuator
		Implements ILightActuator
		Public Sub DoOn() Implements ILightActuator.DoOn
			Console.WriteLine("DoOn")
		End Sub

		Public Sub DoOff() Implements ILightActuator.DoOff
			Console.WriteLine("DoOff")
		End Sub
	End Class

	Partial Public Class Light
'INSTANT VB NOTE: The variable lightActuator was renamed since Visual Basic does not handle variables named the same as their type well:
		Private lightActuator_Renamed As LightActuator
		Private context As LightContext

		Public Sub New()
			Me.lightActuator_Renamed = New LightActuator()
			Me.context = New LightContext(lightActuator_Renamed)
			Me.context.Observer = ObserverConsole.Instance
			Me.context.EnterInitialState()
		End Sub

		Shared Sub Main(ByVal args() As String)
			Dim light_Renamed As New Light()
			light_Renamed.EvOn()
			'light.On();
			'light.On();
			'light.Off();
			Environment.Exit(0)
		End Sub
	End Class
End Namespace

File: Examples/CS/Light/Light.cs
namespace StateForge.Examples.Light
{
    using System;
    using StateForge.StateMachine;

    public interface ILightActuator
    {
        void DoOn();
        void DoOff();
    }

    public class LightActuator : ILightActuator
    {
        public void DoOn()
        {
            Console.WriteLine("DoOn");
        }

        public void DoOff()
        {
            Console.WriteLine("DoOff");
        }
    }

    public partial class Light {
        private LightActuator lightActuator;
        private LightContext context;

        public Light()
        {
            this.lightActuator = new LightActuator();
            this.context = new LightContext(lightActuator);
            this.context.Observer = ObserverConsole.Instance;
            this.context.EnterInitialState();
        }
        
        static void Main(string[] args)
        {
            Light light = new Light();
            light.EvOn();
            //light.On();
            //light.On();
            //light.Off();
            Environment.Exit(0);
        }
    }
}
